src/opensles/AudioInputStreamOpenSLES.cpp

```diff
@@ -124,11 +126,30 @@ Result AudioInputStreamOpenSLES::open() {
     SLresult result = EngineOpenSLES::getInstance().createAudioRecorder(&mObjectInterface,
                                                                         &audioSrc,
                                                                         &audioSink);
+
+    // If the result is "invalid parameter" and we're using the extended data format it could be
+    // that the underlying hardware doesn't support it
+    // (e.g. Moto G: https://github.com/google/oboe/issues/192)
+    // Try again using the old format
+    if (result == SL_RESULT_PARAMETER_INVALID && usingExtendedDataFormat){
+
+        LOGW("Unable to open recording stream using PCM_EX format, trying again using PCM...");
+        audioSink.pFormat = &format_pcm;
+        result = EngineOpenSLES::getInstance().createAudioRecorder(&mObjectInterface,
+                                                                   &audioSrc,
+                                                                   &audioSink);
+        usingExtendedDataFormat = false;
+    }
+
     if (SL_RESULT_SUCCESS != result) {
         LOGE("createAudioRecorder() result:%s", getSLErrStr(result));
         goto error;
     }
 
+    // If the stream is not using the extended data format it must have a format of I16 since no
+    // other format is supported, regardless of what format was requested.
+    if (!usingExtendedDataFormat) mFormat = AudioFormat::I16;
```

Comment:
In AAudio, if you request Format=UNSPECIFIED then you will get either I16 or FLOAT.
If you request FLOAT then you will get a FLOAT stream or the stream will not open.
You do not get an I16 stream. Same for channels.

So we should probably do something like:
    if (formatResuested == UNSPECIFIED) mFormat = I16
    else return ERROR;

Reply:
> or the stream will not open

That's interesting. So it's not like PerformanceMode where if you request LowLatency but don't get it the stream still opens but with PerformanceMode=None.

In that case, like you say, we should make the behaviour the same for OpenSL ES streams. Which is (correct me if I'm wrong)...

For playback streams
==

**API < 21 (only I16 supported)**
Unspecified - Stream opens with format set to I16
Float - Stream does not open and returns relevant error (should we add `ErrorUnsupportedAudioFormat` so it's clear the reason the stream didn't open?)
I16 - Stream opens with format set to I16

**API 21+ (I16 and float supported)**
Unspecified - Stream opens with format set to Float
Float - Stream opens with format set to Float
I16 - Stream opens with format set to I16

For recording streams:
==
**API < 23 (only I16 supported)**
Unspecified - Stream opens with format set to I16
Float - Stream does not open and returns relevant error
I16 - Stream opens with format set to I16

**API 23+ (I16 and float supported)**
Unspecified - Stream opens with format set to Float
Float - Stream opens with format set to Float
I16 - Stream opens with format set to I16

Reply:
Right. If a caller request a specific format then they should either get that format or it should fail.